Add cypress introduction in fr and en (#53)
lunatech-labs/lunatech-blog: posts/2022-06-06-introduction-cypress-fr.adoc

image::accueil-lunatech.png[Page d'accueil du blog de Lunatech, width = 700]

La partie gauche de l'écran va enregistrer toutes les actions effectuées.
La partie de droite montrera en temps réel ce qui se passe sur le site web.

Ne fermez pas cet écran, car le rendu sera actualisé automatiquement au fur et à mesure de vos modifications !

==== b. Vérification du chargement des articles
La seconde étape est de vérifier que nous avons des articles de blog sur la page principale.
Nous n’avons pas accès au code du site, par conséquent, nous utilisons l’inspecteur de notre navigateur pour trouver un moyen de cibler l’élément qui nous intéresse. 
Après avoir vérifié, nous voyons que tous les aperçus d'article possèdent la classe `lb_post`. 
Nous pouvons donc utiliser cet identifiant pour notre exemple.

[source, javascript]
----
describe('First test with Cypress', () => {
 it('should visit', () => {
   cy.visit('https://blog.lunatech.com/')
   cy.get(".lb_post").its('length').should('be.gt', 0)
 })
})
----

Le code se lit très facilement. 
Nous nous attendons à ce qu’il y ait au moins un article sur la page (`gt` correspond à `greater than`). 

> Après avoir sauvegardé, nous voyons le résultat directement dans l’interface graphique :

Reviewer: ```suggestion
Après avoir sauvegardé, le résultat apparait dans l’interface graphique :
```

image::blog-loaded.png[Test vérifiant si les aperçus sont bien affichés, width = 700]

> En bas de la partie gauche, vous pouvez voir `expected 36 to be above 0` comme nous voulions.

Reviewer: ```suggestion
En bas de la partie gauche, vous pouvez voir `expected 36 to be above 0` qui est bien le résultat attendu.
```

Imaginons que pendant la rédaction de notre test, nous nous soyons trompés. 
Au lieu de `greater than`, nous spécifions `lower than` :

[source, javascript]
----
describe('First test with Cypress', () => {
 it('should visit', () => {
   cy.visit('https://blog.lunatech.com/')
   cy.get(".lb_post").its('length').should('be.lt', 0)
 })
})
----

Grâce au rechargement automatique, nous pouvons détecter notre erreur au plus vite avec un message simple et clair : 

image::error.png[Affichage d'une erreur dans notre test, width = 500]

Cypress permet vraiment d’avoir un retour rapide et clair afin de déboguer au plus vite notre code !

==== c. Tri par auteur
Passons à la dernière étape, filtrer par auteur. 
Un peu plus compliqué, mais _pas impossible_ avec cet outil !
Première chose à faire, simuler un clic de souris sur un auteur afin d’activer le tri. 

> Nous allons me prendre comme exemple (Vincent Brulé). 

Reviewer: ```suggestion
Pour cet exemple, l'auteur sera moi-même (Vincent Brulé). 
```

Nous devons donc trouver un élément correspondant à mon nom et cliquez dessus. 
L'élément qui nous intéresse est le suivant :

image::author.png[Élément représentant un auteur, width = 200]

Il y a plusieurs façons de procéder. 

> Pour l’exemple, nous allons chercher un élément avec le texte *Vincent Brulé* sans passer par une classe ou un identifiant.

Reviewer: Il y a beaucoup de fois le terme `pour l'exemple`, peut etre ici c'est moins necessaire.
```suggestion
Commençons par essayer de chercher un élément avec le texte *Vincent Brulé* sans passer par une classe ou un identifiant.
```
par exemple

   cy.contains('Vincent Brulé').click()
   // We check the content of the url
   cy.url().should('include', 'VincentBrule')
   // We check that the author of all the previews in the page is Vincent Brulé
   cy.get(".lb_post-author").each((author) => {
      expect(author.text()).to.equal('Vincent Brulé')
   })
 })
})
----

On sauvegarde et c’est tout, notre exemple est déjà implémenté ! 
Au passage, une fonctionnalité très utile de l’outil graphique est de pouvoir explorer les étapes et voir directement sur la partie droite les actions effectuées par Cypress. 
Par exemple, à la fin de notre exemple, nous voyons la page avec seulement mes articles :

image::last-example.png[Résultat de notre exemple, width = 700]

Et si on passe la souris sur la première étape où nous avons sélectionné tous les articles, on voit dans le rendu la page d’accueil avec tous les éléments en bleu : 

image::replay.png[Nous pouvons revoir toutes les étapes une par une, width = 700]

C’est très utile pour déboguer et on peut aussi enregistrer une vidéo de toutes ces actions afin de déboguer plus tard par exemple !

Nous sommes arrivés au bout de ce premier article à propos de Cypress.

> J’espère que je vous ai donné envie d’en savoir plus, car nous allons explorer en détails le fonctionnement de Cypress et les bonnes pratiques dans l’article suivant !

Reviewer: ```suggestion
J’espère que je vous ai donné envie d’en savoir plus, car nous allons explorer en détail le fonctionnement de Cypress et les bonnes pratiques dans l’article suivant !
```
